### PR TITLE
ricochet-refresh: init at 3.0.10

### PR DESCRIPTION
**Pull request by morganava**
###### Motivation for this change
Initial package for Ricochet Refresh v3.0.10 ( https://github.com/blueprint-freespeech/ricochet-refresh/releases/tag/v3.0.10-release )

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 128440` at add61889 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>ricochet-refresh</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:52:3:
```
   |
52 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -1370,6 +1370,12 @@
     github = "bloomvdomino";
     githubId = 33204710;
   };
+  blueprint-freespeech = {
+    email = "ricochet@blueprintforfreespeech.net";
+    github = "blueprint-freespeech";
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:1375]: Your github handle is pospeselr.

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+, fetchurl
+, fontconfig
+, freetype
+, libxcb
+, libxkbcommon
+, xorg
+}:
+
+assert stdenv.hostPlatform.system == "x86_64-linux";
+
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:13]:
```suggestion
```
use platforms instead.

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+let
+  version = "3.0.10";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:15]:
```suggestion
```

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    name = "ricochet-refresh";
+    exec = "ricochet-refresh";
+    icon = "ricochet-refresh";
+    desktopName = "Ricochet Refresh";
+    genericName = "Ricochet Refresh";
+    categories = "Network;InstantMessaging;";
+  };
+
+in stdenv.mkDerivation {
+  name = "ricochet-refresh-${version}";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:39]:
```suggestion
in stdenv.mkDerivation rec {
  pname = "ricochet-refresh";
  version = "3.0.10";
```

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+  # nix by default strips binaries (to save space?) but there is
+  # an outstanding issue where this breaks if rpath is set to a longer value
+  # using patchelf prior
+  # see: https://github.com/NixOS/patchelf/issues/10
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:44]:
```suggestion
  # stripping breaks if rpath is set to a longer value using patchelf prior
  # see: https://github.com/NixOS/patchelf/issues/10
```
There is no reason you do not want to have an unstripped binary in a production system.

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  dontStrip = true;
+
+  src = fetchurl {
+    url = "https://github.com/blueprint-freespeech/ricochet-refresh/releases/download/v${version}-release/ricochet-refresh-${version}-linux-x86_64.tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:48]: Can we use the github tarball instead?

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    sha256 = "32eca517e0d20e5b427c6d9beafa4f18373b7417907c89e01beebe3934caf32e";
+  };
+
+  installPhase = ''
+    echo "unpacking $src..."
+    tar -xvf $src
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:54]: Belongs into unpackPhase.

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+    # deploy icon
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:56]:
```suggestion
```

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    mkdir -p $out/share/pixmaps
+    mv ricochet-refresh/ricochet-refresh.png $out/share/pixmaps/ricochet-refresh.png
+
+    # deploy bins
+    mkdir -p $out
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:61]:
```suggestion
```
already exists

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    mv ricochet-refresh $out/bin
+
+    # patch bins
+    pushd $out/bin
+    for file in ricochet-refresh tor; do
+      echo "patching $file..."
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:67]:
```suggestion
```

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+    # deploy icon
+    mkdir -p $out/share/pixmaps
+    mv ricochet-refresh/ricochet-refresh.png $out/share/pixmaps/ricochet-refresh.png
+
+    # deploy bins
+    mkdir -p $out
+    mv ricochet-refresh $out/bin
+
+    # patch bins
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:64]:
```suggestion
```

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    libxkbcommon
+    stdenv.cc.cc.lib
+    xorg.xcbutilimage
+    xorg.xcbutilkeysyms
+    xorg.xcbutilrenderutil
+    xorg.xcbutilwm
+    xorg.libX11
+    xorg.libXext
+  ];
+  desktopItem = makeDesktopItem {
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:29]: Why not use copyDesktopItems?

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    pushd $out/bin
+    for file in ricochet-refresh tor; do
+      echo "patching $file..."
+      patchelf --set-interpreter ${stdenv.cc.bintools.dynamicLinker} --set-rpath ${rpath} $file
+    done
+    popd
+
+    # deploy desktop entry data
+    mkdir -p "$out/share/applications"
+    cp ${desktopItem}/share/applications/* $out/share/applications
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:74]: redundant with copyDesktopItems

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+  '';
+
+  meta = {
+    description = "Private, anonymous, and metadata resistant instant messaging using Tor onion services";
+    homepage = "https://www.ricochetrefresh.net";
+    license = lib.licenses.bsd3;
+    platforms = [ "x86_64-linux" ];
+    maintainers = [ lib.maintainers.blueprint-freespeech ];
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:82]:
```suggestion
  meta = with lib; {
    description = "Private, anonymous, and metadata resistant instant messaging using Tor onion services";
    homepage = "https://www.ricochetrefresh.net";
    license = licenses.bsd3;
    platforms = [ "x86_64-linux" ];
    maintainers = [ maintainers.blueprint-freespeech ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (pospeselr) [reply maintainers/maintainer-list.nix:1375]: Ricochet-Refresh is maintained by Blueprint for Freespech, ricochet@blueprintforfreespeech.net goes to whomever the current developer maintaining it is.

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix (hunk 6 above)
Comment (pospeselr) [reply pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:37]: No, this is a binary release

**Comment by pospeselr** [comment]
Ok I believe that is all of your feedback resolved. It took quite a bit of finagling to get the copyDesktopItem script to run properly. Your documentation should probably be updated to include how to use it.

In general, it's a bit hard to figure out what the current best-practices are for new packages.

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (SuperSandro2000) [reply maintainers/maintainer-list.nix:1381]: Orgs are not supposed to be added to the maintainer-list.nix. Please create a team for them instead. maintainer-list.nix is for single accounts.

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix (hunk 6 above)
Comment (SuperSandro2000) [reply pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:37]: Is there a reason why we are not building from source here?

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+    # patch bins
+    pushd $out/bin
+    for file in ricochet-refresh tor; do
+      patchelf --set-interpreter ${stdenv.cc.bintools.dynamicLinker} --set-rpath ${rpath} $file
+    done
+    popd
+
+    # trigger running of copyDesktop Items
+    runHook postInstall
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:73]:
```suggestion
    runHook postInstall
```

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+  ];
+  nativeBuildInputs = [ copyDesktopItems ];
+
+  unpackPhase = ''
+    echo "unpacking $src..."
+    tar -xvf $src
+  '';
+
+  installPhase = ''
+    # deploy icon
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:58]:
```suggestion
  installPhase = ''
    runHook preInstall

    # deploy icon
```

**Review by SuperSandro2000 — COMMENTED** [review]
hooks are always supposed to be run, they do not require any documentation.

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix (hunk 15 above)
Comment (pospeselr) [reply pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:73]: I had to add the postInstall hook or else the generated desktop file would not be deployed. As it stands, I still need to logout/login for Ricochet to show up in the application list.

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix (hunk 16 above)
Comment (pospeselr) [reply pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:58]: It's nice to split up blocks of code with section headers (even if they are technically redundant to the one-liner that follows) to give future maintainers a better chance to understand wth is going on.

Comment (pospeselr) [reply pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:58]: What does the preInstall hook here do?

**Comment by pospeselr** [comment]
Switched to using a team list and added the preInstall hook

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix (hunk 6 above)
Comment (pospeselr) [reply pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:37]: The build system/scripts aren't exactly stable, binary release makes it easier for us to maintain.

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+  # see: https://github.com/NixOS/patchelf/issues/10
+  dontStrip = true;
+
+  src = fetchurl {
+    url = "https://github.com/blueprint-freespeech/ricochet-refresh/releases/download/v${version}-release/ricochet-refresh-${version}-linux-x86_64.tar.gz";
+    sha256 = "32eca517e0d20e5b427c6d9beafa4f18373b7417907c89e01beebe3934caf32e";
+  };
+
+  desktopItems = [ (makeDesktopItem {
+      name = "ricochet-refresh";
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:42]:
```suggestion
  desktopItems = [
    (makeDesktopItem {
      name = "ricochet-refresh";
```

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+
+    # patch bins
+    pushd $out/bin
+    for file in ricochet-refresh tor; do
+      patchelf --set-interpreter ${stdenv.cc.bintools.dynamicLinker} --set-rpath ${rpath} $file
+    done
+    popd
+
+    # trigger running of copyDesktop Items
+    runHook postInstall
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:75]:
```suggestion
    runHook preInstall

    mkdir -p $out/share/pixmaps
    mv ricochet-refresh/ricochet-refresh.png $out/share/pixmaps/ricochet-refresh.png
    mv ricochet-refresh $out/bin

    pushd $out/bin
    for file in ricochet-refresh tor; do
      patchelf --set-interpreter ${stdenv.cc.bintools.dynamicLinker} --set-rpath ${rpath} $file
    done
    popd

    runHook postInstall
```

**Comment by milahu** [comment]
<details>
<summary>let me resurrect this review thread</summary>

> pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix
> ```
>   dontStrip = true;
> 
>   src = fetchurl {
>     url = "https://github.com/blueprint-freespeech/ricochet-refresh/releases/download/v${version}-release/ricochet-refresh-${version}-linux-x86_64.tar.gz";
> ```
>
> @SuperSandro2000 on Jun 30 Member
> Can we use the github tarball instead?

> @pospeselr pospeselr on Jul 11 Author
> No, this is a binary release

> @SuperSandro2000 SuperSandro2000 on Jul 11 Member
> Is there a reason why we are not building from source here?

> @pospeselr pospeselr on Jul 23 Author
> The build system/scripts aren't exactly stable, binary release makes it easier for us to maintain.

</details>

> binary release makes it easier for us to maintain

binary release makes it easier to include spyware ...

> The build system/scripts aren't exactly stable

then lets fix the build system. it worked with [the old ricochet](https://github.com/ricochet-im/ricochet/blob/master/BUILDING.md), no?

related:
https://github.com/blueprint-freespeech/ricochet-refresh/issues/104
https://github.com/blueprint-freespeech/ricochet-refresh/issues/110

File: pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix (hunk 6 above)
Comment (SuperSandro2000) [reply pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:37]: According to their docs https://github.com/blueprint-freespeech/ricochet-refresh/blob/main/BUILDING.md#qt-sdk it looks like plain standard qmake.

Comment (milahu) [reply pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:37]: > looks like plain standard qmake

yepp, see my [ninja edit](https://github.com/NixOS/nixpkgs/pull/128440#issuecomment-903491817)

Comment (milahu) [reply pkgs/applications/networking/instant-messengers/ricochet-refresh/default.nix:37]: @pospeselr fixed : )

**Comment by m-simonelli** [comment]
Hi, we're currently migrating to CMake, so these scripts will soon be out of date. See [this](https://github.com/blueprint-freespeech/ricochet-refresh/pull/118). I'll update with the PR for the new build system soon, but building should be as simple as `mkdir build && cd build && cmake -S ../src && cmake --build .` :)

**Comment by SuperSandro2000** [comment]
> Hi, we're currently migrating to CMake, so these scripts will soon be out of date. See [this](https://github.com/blueprint-freespeech/ricochet-refresh/pull/118). I'll update with the PR for the new build system soon, but building should be as simple as `mkdir build && cd build && cmake -S ../src && cmake --build .` :)

cmake should do everything via its wrapper in NixOS.

**Comment by AndersonTorres** [comment]
Closing as dead and deprecated

**Pull request closed without merging**